src/openforms/registrations/api/views.py

```diff
@@ -86,13 +123,48 @@ def get_objects(self):
             catalogus["url"]: catalogus for catalogus in catalogus_data
         }
 
-        iotypen_data = client.get_all_informatieobjecttypen()
-        iotypen = [
-            {
-                "informatieobjecttype": factory(InformatieObjectType, iotype),
-                "catalogus": factory(Catalogus, catalogus_mapping[iotype["catalogus"]]),
-            }
-            for iotype in iotypen_data
-        ]
+        if "catalogus_domein" in filter_serializer.validated_data:
+            domein = filter_serializer.validated_data["catalogus_domein"]
+            # `catalogus_rsin` is guaranteed to be present if `catalogus_domein` is:
+            rsin = filter_serializer.validated_data["catalogus_rsin"]
+
+            matching_catalogus: str | None = next(
+                (
+                    catalogus["url"]
+                    for catalogus in catalogus_data
+                    if catalogus["domein"] == domein and catalogus["rsin"] == rsin
+                ),
+                None,
+            )
+
+            if matching_catalogus is None:
+                return []
+
+            iotypen_data = client.get_all_informatieobjecttypen(
+                catalogus=matching_catalogus
+            )
+        else:
+            iotypen_data = client.get_all_informatieobjecttypen()
+
+        iotypen = []
+
+        for iotype in iotypen_data:
+            # fmt: off
+            exists = any(
+                existing_iotype["informatieobjecttype"].omschrijving == iotype["omschrijving"]
+                and existing_iotype["informatieobjecttype"].catalogus == iotype["catalogus"]
+                for existing_iotype in iotypen
+            )
+            # fmt: on
+
+            if not exists:
+                iotypen.append(
+                    {
+                        "informatieobjecttype": factory(InformatieObjectType, iotype),
```

Comment:
I think you want to use a different data model for `InformatieObjectType` since now this still includes a URL and version data, and we're basically only interested in the combination of `(domein, rsin, iot_omschrijving)` tuples, so let's just make the data non-confusing and discard what isn't relevant.

Probably that way you can also use a nicer way to build the collection of unique document types, maybe a set of tuples and then you sort it based on catalogus.rsin / catalogus.domein / iot.omschrijving for a stable ordering.